src/vsphere_cpi/spec/unit/cloud/vsphere/file_provider_spec.rb

```diff
@@ -2,86 +2,194 @@
+            {'Cookie' => "vmware_cgi_ticket=ticket"}
+          ).exactly(Retryer::MAX_TRIES).times
+           .and_return(double('response', code: 404))
+
           expect(http_client).to receive(:get).with(
             'https://fake-vcenter-host/folder/fake-path?'\
-            'dcPath=fake-datacenter-name%201&dsName=fake-datastore-name%201', {}
+          'dcPath=fake-datacenter-name%201&dsName=fake-datastore-name%201', {}
           ).and_return(double('response', code: 404))
 
           expect(
-            file_provider.fetch_file_from_datastore(datacenter_name, datastore_name, path)
+            file_provider.fetch_file_from_datastore(datacenter_name, datastore, path)
           ).to be_nil
         end
       end
 
       context 'when vSphere cannot handle the request' do
         it 'retries then raises an error' do
-          expect(http_client).to receive(:get)
-            .with(
-              'https://fake-vcenter-host/folder/fake-path?'\
-              'dcPath=fake-datacenter-name%201&dsName=fake-datastore-name%201', {}
-            )
-            .exactly(Retryer::MAX_TRIES).times
-            .and_return(double('response', code: 500))
+          expect(http_client).to receive(:get).with(
+            'https://host/folder/fake-path?'\
+            'dsName=fake-datastore-name%201',
+            {'Cookie' => "vmware_cgi_ticket=ticket"}
+          ).exactly(Retryer::MAX_TRIES).times
+           .and_return(double('response', code: 500))
+
+          expect(http_client).to receive(:get).with(
+            'https://fake-vcenter-host/folder/fake-path?'\
+          'dcPath=fake-datacenter-name%201&dsName=fake-datastore-name%201', {}
+          ).exactly(Retryer::MAX_TRIES).times
+           .and_return(double('response', code: 500))
 
           expect {
-            file_provider.fetch_file_from_datastore(datacenter_name, datastore_name, path)
+            file_provider.fetch_file_from_datastore(datacenter_name, datastore, path)
           }.to raise_error(/Could not transfer file/)
         end
       end
     end
 
     describe '#upload_file_to_datastore' do
-      let(:upload_contents) {"fake uplóad contents"}
-      it 'uploads specified file' do
+      let(:upload_contents) {"fake upload contents"}
+      it 'first tries to upload file to datastore via a host and succeeds' do
         response = double('response', code: 200, body: nil)
+        expect(http_client).to receive(:put).with(
+          'https://host/folder/fake-path?'\
+          'dsName=fake-datastore-name%201',
+          upload_contents,
+          { 'Content-Type' => 'application/octet-stream', 'Content-Length' => upload_contents.bytesize,
+            'Cookie' => "vmware_cgi_ticket=ticket"
+          }
+        ).and_return(response)
+
+        file_provider.upload_file_to_datastore(datacenter_name, datastore, path, upload_contents)
+      end
+
+      it 'fails to upload file via host and retries to upload through datacenter and uploads specified file' do
```

Comment:
fails to upload file via host and retries to upload through datacenter and succeeds